src/portgraph.rs

```diff
@@ -871,40 +843,38 @@ impl PortGraph {
         }
 
-        let port_list = if incoming + outgoing > 0 {
-            let new_port_list = self.alloc_ports(node, incoming, outgoing);
-            if let Some(old_port_list) = old_port_list {
-                let incoming_rekeys = (0..old_incoming).zip(0..incoming);
-                let outgoing_rekeys = (old_incoming..old_total).zip(incoming..new_total);
-                for (old, new) in incoming_rekeys.chain(outgoing_rekeys) {
-                    let old_index = old_port_list.index() + old;
-                    let new_index = new_port_list.index() + new;
-                    let old_port = PortIndex::new(old_index);
-                    let new_port = PortIndex::new(new_index);
-
-                    if let Some(link) = self.port_link[old_index] {
-                        self.port_link[link.index()] = Some(new_port);
-                    }
-                    self.port_link[new_index] = self.port_link[old_index].take();
-                    self.port_meta[new_index] = self.port_meta[old_index];
-
-                    rekey(old_port, Some(new_port));
+        if 0 < new_total && new_total <= old_capacity {
+            // Special case when we can avoid reallocations by shifting the
+            // ports in the pre-allocated slab.
+            self.resize_ports_inplace(node, incoming, outgoing, rekey);
+            return;
+        }
+        // Allocate a new slab of ports. If `new_total` is 0, we just free the
+        // old slab.
+        //
+        // TODO: Fine-tune overallocation factor
+        let target_capacity = new_total.next_power_of_two();
+        let new_meta = self.alloc_ports(node, incoming, outgoing, target_capacity - new_total);
+
+        // Move the port data.
+        for dir in Direction::BOTH {
+            let rekeys = node_meta.ports(dir).zip(new_meta.ports(dir));
+            for (old, new) in rekeys {
+                if let Some(link) = self.port_link[old] {
+                    self.port_link[link.index()] = Some(PortIndex::new(new));
                 }
-                self.free_ports(old_port_list, old_incoming + old_outgoing);
-            }
-            Some(new_port_list)
-        } else {
-            if let Some(plist) = old_port_list {
-                self.free_ports(plist, old_incoming + old_outgoing);
+                self.port_link[new] = self.port_link[old].take();
+                self.port_meta[new] = self.port_meta[old];
+
+                rekey(PortIndex::new(old), Some(PortIndex::new(new)));
             }
-            None
-        };
+        }
 
-        self.node_meta[node.index()] =
-            NodeEntry::Node(NodeMeta::new(port_list, incoming as u16, outgoing as u16));
+        self.node_meta[node.index()] = NodeEntry::Node(new_meta);
+        self.free_ports(old_port_list, old_capacity);
 
-        self.port_count -= old_incoming + old_outgoing;
-        self.port_count += incoming + outgoing;
+        self.port_count -= old_total;
+        self.port_count += new_total;
```

Comment:
```suggestion
        self.port_count += new_total - old_total;
```

Reply:
This needs some care to avoid underflows